### ISSUE TITLE
Adding utility classes/objects like Vector and Point

### ISSUE THREAD

**Issue by camchenry** [davisdude/mlib#9]
I came here by way of the post about Hacktober on love2d forums. I have never used this library before, but it looks quite interesting. You said you were thinking about adding vectors, so I am making this issue to start some discussion on if/how they should be implemented.

**Comment by davisdude** [comment]
Great! Thanks for the interest!

If objects are used, I think they would have to be accepted as arguments by all functions. 
There might be some conflict of when when you're passing objects vs when you're passing variables to the function, which may cause some trouble. 

My idea is that there would be a simple Lua class that creates all these objects. Something like: 
```Lua
local function isObject( obj ) 
    return obj.__type and obj.__isMlibObject
end

local classes = {
    point = { -- Name of class
        new = function( x, y ) -- Creation function
            -- You could also add type checking here, etc.
            return { x = x, y = y, __type = 'point', __isMlibObject = true }
        end, 
        checkCircle = function( self, ... )
            local input = checkInput( ... )
            local x, y, radius
            if #input == 1 and input.__type == 'circle' then
                x, y, radius = point.x, point.y, point.y
            elseif #input == 3 then
                x, y, radius = unpack( input )
            end
            return circleCheckPoint( self.x, self.y, x, y, radius )
        end, 
        -- etc.
    }
    -- etc.
}

for i, v in pairs( classes ) do
    setmetatable( v, { __call = function( t, ... ) return v.new( ... ) } )
end

local function createClass( type, ... )
    return classes[type]( ... )
end
```
This isn't tested, but should be a good baseline.

**Comment by camchenry** [comment]
I agree, it seems really clunky to have all of the functions check if the first argument is an object or a number. That code does seem like it would use a lot of memory, but I haven't thoroughly used metatables. So, I might mess around with that and see if I could get it working.

**Comment by karai17** [comment]
Metatables don't use any extra memory. Tables are passed by reference, not copied, so using a single table as the meta table for 1000 other tables only gives you the footprint of 1001 tables, not 2000.

**Comment by davisdude** [comment]
Metatables definitely seem like the way to go. What I was doing (above) doesn't really seem  suited for this.

**Comment by camchenry** [comment]
Looking at #10, should the point code match the vector code?

**Comment by karai17** [comment]
What exactly do you mean by point? A vector can easily be described as a point already.

**Comment by camchenry** [comment]
Fair enough, that's sort of what I was thinking anyway. Unless there are other specific objects that should be added I think this issue can be closed.

**Comment by davisdude** [comment]
Well, the current vector implementation isn't really an object, so it's actually going more for "dick typing" on that one.

**Comment by karai17** [comment]
Depending on the scope of this library, there are lots of potential objects that could be added such as matrices of varying sizes, vec3s, etc. mlib has a very different structure to it than my own math library, CPML, which contains many of the same features and isolates modules in a very intuitive and friendly way.

**Comment by davisdude** [comment]
Yeah, due to the way I implemented some functions it was impossible to make it so that files were separated (i.e polygon.lua, etc.) without some weirdness occurring.

**Comment by karai17** [comment]
Well you don't necessarily need to separate files the same way that we do.
It is helpful, though, since our library references other modules within,
but one large file can do the same thing.

Requiring cpml gives you a table with access to all modules, but you can
require only the modules you need instead

```lua
local cpml = require "cpml"
local vector = cpml.vec3(0, 0, 0)
```

```lua
local vec3 = require"cpml.modules.vec3"
local vector = vec3(0, 0, 0)
```


---
Landon Manning
lmanning17@gmail.com

**Comment by karai17** [comment]
God dammit GitHub.

**Comment by davisdude** [comment]
I think it needs to be `Lua`,  not `lua`

**Comment by karai17** [comment]
Nah, it's an issue with email responses.

**Comment by davisdude** [comment]
Oh, okay.
